File: orig_app/internal/handler/snippets/snippets_actions.go
package snippets

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/handler/blob"
	"gitlab.com/gitlab-org/gitlab-redux/internal/handler/notes"
	"gitlab.com/gitlab-org/gitlab-redux/internal/handler/pagination"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
)

// SnippetsActions provides actions for handling snippets
type SnippetsActions struct {
	rendersNotes      *notes.RendersNotes
	rendersBlob       *blob.RendersBlob
	paginatedCollection *pagination.PaginatedCollection
	noteableMetadata  *service.NoteableMetadataService
	sendBlob          *SendBlob
	snippetsSort      *SnippetsSort
	analyticsTracking *service.AnalyticsTrackingService
	workhorseService  *service.WorkhorseService
}

// NewSnippetsActions creates a new instance of SnippetsActions
func NewSnippetsActions(
	rendersNotes *notes.RendersNotes,
	rendersBlob *blob.RendersBlob,
	paginatedCollection *pagination.PaginatedCollection,
	noteableMetadata *service.NoteableMetadataService,
	sendBlob *SendBlob,
	snippetsSort *SnippetsSort,
	analyticsTracking *service.AnalyticsTrackingService,
	workhorseService *service.WorkhorseService,
) *SnippetsActions {
	return &SnippetsActions{
		rendersNotes:      rendersNotes,
		rendersBlob:       rendersBlob,
		paginatedCollection: paginatedCollection,
		noteableMetadata:  noteableMetadata,
		sendBlob:          sendBlob,
		snippetsSort:      snippetsSort,
		analyticsTracking: analyticsTracking,
		workhorseService:  workhorseService,
	}
}

// RegisterRoutes registers the routes for snippets actions
func (s *SnippetsActions) RegisterRoutes(router *gin.Engine) {
	// Register routes for snippets actions
	// In Go/Gin, we would typically use router groups
	snippetsGroup := router.Group("/snippets")
	{
		snippetsGroup.GET("/:id/edit", s.Edit)
		snippetsGroup.GET("/:id/raw", s.Raw)
		snippetsGroup.GET("/:id", s.Show)
	}
}

// Edit handles the edit action for a snippet
func (s *SnippetsActions) Edit(c *gin.Context) {
	// In Go, we would typically render a template or return JSON
	// This is a placeholder for the actual implementation
	c.HTML(http.StatusOK, "snippets/edit", gin.H{
		"snippet": c.MustGet("snippet"),
	})
}

// Raw handles the raw action for a snippet
// This endpoint is being replaced by Snippets::BlobController#raw
// Support for old raw links will be maintained via this action but
// it will only return the first blob found
func (s *SnippetsActions) Raw(c *gin.Context) {
	snippet := c.MustGet("snippet").(*model.Snippet)
	blob := s.getBlob(snippet)

	// Set content type for workhorse
	s.workhorseService.SetContentType(c)

	// Check if the blob has a snippet (old format)
	if blob.Snippet != nil {
		// Send the data with appropriate headers
		content := s.convertLineEndings(blob.Data, c.Query("line_ending"))
		filename := model.SanitizeFileName(blob.Name)

		c.Header("Content-Type", "text/plain; charset=utf-8")
		c.Header("Content-Disposition", "attachment; filename=\""+filename+"\"")
		c.String(http.StatusOK, content)
	} else {
		// Send the snippet blob using the SendBlob service
		s.sendBlob.SendSnippetBlob(c, snippet, blob)
	}
}

// Show handles the show action for a snippet
func (s *SnippetsActions) Show(c *gin.Context) {
	snippet := c.MustGet("snippet").(*model.Snippet)

	// Track analytics event
	s.analyticsTracking.TrackEvent(c, "i_snippets_show", nil)

	// Handle different formats
	format := c.GetHeader("Accept")
	if strings.Contains(format, "application/javascript") || c.Query("format") == "js" {
		s.showJS(c, snippet)
	} else {
		s.showHTML(c, snippet)
	}
}

// IsJSRequest checks if the request is a JavaScript request
func (s *SnippetsActions) IsJSRequest(c *gin.Context) bool {
	format := c.GetHeader("Accept")
	return strings.Contains(format, "application/javascript") || c.Query("format") == "js"
}

// Private helper methods

func (s *SnippetsActions) showHTML(c *gin.Context, snippet *model.Snippet) {
	// Create a new note for the snippet
	note := &model.Note{
		Noteable: snippet,
		Project:  snippet.Project,
	}

	// Get discussions and notes
	discussions := snippet.Discussions
	notes := s.prepareNotesForRendering(discussions)

	// Render the template
	c.HTML(http.StatusOK, "snippets/show", gin.H{
		"snippet":     snippet,
		"note":        note,
		"noteable":    snippet,
		"discussions": discussions,
		"notes":       notes,
	})
}

func (s *SnippetsActions) showJS(c *gin.Context, snippet *model.Snippet) {
	if !snippet.IsEmbeddable() {
		c.Status(http.StatusNotFound)
		return
	}

	// Conditionally expand blobs
	blobs := s.getBlobs(snippet)
	s.conditionallyExpandBlobs(blobs)

	// Render the template
	c.HTML(http.StatusOK, "shared/snippets/show", gin.H{
		"snippet": snippet,
		"blobs":   blobs,
	})
}

func (s *SnippetsActions) getBlob(snippet *model.Snippet) *model.Blob {
	blobs := s.getBlobs(snippet)
	if len(blobs) > 0 {
		return blobs[0]
	}
	return nil
}

func (s *SnippetsActions) getBlobs(snippet *model.Snippet) []*model.Blob {
	if snippet.IsEmptyRepo() {
		return []*model.Blob{snippet.Blob}
	}
	return snippet.Blobs
}

func (s *SnippetsActions) convertLineEndings(content string, lineEnding string) string {
	if lineEnding == "raw" {
		return content
	}
	return strings.ReplaceAll(content, "\r\n", "\n")
}

func (s *SnippetsActions) prepareNotesForRendering(discussions []*model.Discussion) []*model.Note {
	var notes []*model.Note
	for _, discussion := range discussions {
		notes = append(notes, discussion.Notes...)
	}
	return s.rendersNotes.PrepareNotesForRendering(notes)
}

func (s *SnippetsActions) conditionallyExpandBlobs(blobs []*model.Blob) {
	// Implementation would depend on how blobs are expanded
	// This is a placeholder for the actual implementation
}

File: rust_app/src/controllers/concerns/snippets_actions.rs
// Ported from orig_app/app/controllers/concerns/snippets_actions.rb on 2025-04-29
// This trait provides actions for handling snippets, ported from the Ruby concern.
use actix_web::{web, HttpResponse, Responder};
use serde::Deserialize;

use crate::controllers::concerns::{
    NoteableMetadata, PaginatedCollection, ProductAnalyticsTracking, RendersBlob, RendersNotes,
    SnippetsSort,
};
use crate::models::blob::Blob;
use crate::models::discussion::Discussion;
use crate::models::note::Note;
use crate::models::snippet::Snippet;

pub trait SnippetsActions:
    RendersNotes
    + RendersBlob
    + PaginatedCollection
    + NoteableMetadata
    + SnippetsSort
    + ProductAnalyticsTracking
{
    fn snippet(&self) -> &Snippet;
    fn set_snippet(&mut self, snippet: Snippet);
    fn is_js_request(&self) -> bool;
    fn workhorse_set_content_type(&self);
    fn convert_line_endings(&self, content: &str, line_ending: &str) -> String {
        // Ported from Ruby: params[:line_ending] == 'raw' ? content : content.gsub(/\r\n/, "\n")
        if line_ending == "raw" {
            content.to_string()
        } else {
            content.replace("\r\n", "\n")
        }
    }
    fn send_snippet_blob(&self, snippet: &Snippet, blob: &Blob) -> HttpResponse;
    fn sanitized_file_name(&self, name: &str) -> String;
    fn content_disposition(&self) -> String;

    fn edit(&self) -> impl Responder {
        // In a real implementation, this would render the edit view
        HttpResponse::Ok().finish()
    }

    fn raw(&self) -> impl Responder {
        self.workhorse_set_content_type();
        let blob = self.blob();
        // In Ruby: if defined?(blob.snippet)
        if blob.has_snippet() {
            let data = self.convert_line_endings(&blob.data, "raw");
            let filename = self.sanitized_file_name(&blob.name);
            HttpResponse::Ok()
                .content_type("text/plain; charset=utf-8")
                .header("Content-Disposition", self.content_disposition())
                .body(data)
        } else {
            self.send_snippet_blob(self.snippet(), blob)
        }
    }

    fn show(&self) -> impl Responder {
        if self.is_js_request() {
            if self.snippet().embeddable() {
                self.conditionally_expand_blobs(self.blobs());
                // Would render 'shared/snippets/show' in a real implementation
                HttpResponse::Ok().finish()
            } else {
                HttpResponse::NotFound().finish()
            }
        } else {
            let note = Note::new(self.snippet(), self.snippet().project());
            let discussions = self.snippet().discussions();
            let notes =
                self.prepare_notes_for_rendering(discussions.iter().flat_map(|d| d.notes()));
            // Would render 'show' in a real implementation
            HttpResponse::Ok().finish()
        }
    }

    fn blob(&self) -> &Blob {
        self.blobs().first().expect("No blobs found")
    }

    fn blobs(&self) -> Vec<&Blob> {
        if self.snippet().empty_repo() {
            vec![self.snippet().blob()]
        } else {
            self.snippet().blobs()
        }
    }
}
